doc/architectural/goto-program-transformations.md

```diff
@@ -0,0 +1,249 @@
+`__CPROVER_mm_io_r` and `__CPROVER_mm_io_w` modelling functions if they exist.
+See [modeling-mmio.md] for details of modeling memory-mapped I/O regions of
+device interfaces. This pass is always carried out but will only make changes if
+one of the modelling functions exist.
+
+_Predecessor pass is [Removal/Lowering of Function Pointers]._
+
+
+### Instrument/Remove Preconditions
+
+This pass moves preconditions associated with functions to call sites rather
+than at the head of the function. Note that functions may have multiple call
+sites and in this case the instrumentation is copied to these multiple call
+sites.
+
+_Predecessor pass is [Memory Mapped IO Instrumentation]._
+
+
+### Removal/Lowering of Return Statements
+
+This pass replaces returns of values with writes and reads to global variables.
+
+_Predecessor passes are [Instrument/Remove Preconditions] or [Partial Inlining]._
+
+
+### Remove/Lower Vector Typed Expressions
+
+This pass removes/lowers expressions corresponding to CPU instruction sets such
+as MMX, SSE and AVX. For more details on how this is done see vector_typet and
+remove_vector.cpp.
+
+_Predecessor pass is [Removal/Lowering of Return Statements]._
+
+
+### Remove/Lower Complex Typed Expressions
+
+This pass is for the handling of complex numbers in the mathematical sense of a
+number consisting of paired real and imaginary parts. These are removed/lowered
+in this pass.
+
+_Predecessor pass is [Remove/Lower Vector Typed Expressions]._
+
+
+### Rewrite Unions
+
+This pass converts union member access into byte extract and byte update
+operations. 
+
+**TODO: Pointer/type related operations require closer implementation
+examination.**
+
+_Predecessor pass is [Remove/Lower Complex Typed Expressions]._
+
+
+### goto_check_c
+
+This is a pass that can add many checks and instrumentation largely related to
+the definition of the C language. Note that this transformation pass is
+mandatory in the current implementation, however it may have no effect depending
+on the command line options specified.
+
+**TODO: add some more details/hints on what the options of note are**
+
+_Predecessor pass is [Rewrite Unions]._
+
+
+### Adjust Float Expressions
+
+This is a transform from general mathematical operations over floating point
+types into floating point specific operator variations which include a rounding
+mode.
+
+_Predecessor passes are [goto_check_c] or [Transform Assertions Assumptions]._
+
+
+### Goto Functions Update
+
+This transformation updates **(in memory?)** data goto program fields which may
+have been invalidated by previous passes. Note that the following are performed
+by this pass:
+- Incoming edges
+- Target numbers
+- location numbers
+- loop numbers
+
+_Predecessor passes are [Adjust Float Expressions] or [String Abstraction]._
+
+
+### Add Failed Symbols
+
+This pass adds failed symbols to the symbol table only. See
+`src/pointer-analysis/add_failed_symbols.h` for details.
```

Comment:
What is a failed symbol, why/how is it used? the source code does not contain a high-level description of this.

Reply:
They are explained here - https://github.com/diffblue/cbmc/blob/d7099aeaff5043b65aba669fe221cc9b9af40c8c/src/goto-symex/symex_dereference.cpp#L457

My reading of this is that these symbols are used as an (undefined) resulting value of a pointer dereference in the case where a pointer may point to a target which is not defined by the model.